boa/src/builtins/object/mod.rs

```diff
@@ -490,58 +226,168 @@ impl Object {
     /// - [EcmaScript reference][spec]
     ///
     /// [spec]: https://tc39.es/ecma262/#sec-isconstructor
+    #[inline]
     pub fn is_constructable(&self) -> bool {
-        match self.func {
-            Some(ref function) => function.is_constructable(),
-            None => false,
+        matches!(self.data, ObjectData::Function(ref f) if f.is_constructable())
+    }
+
+    /// Checks if it an `Array` object.
+    #[inline]
+    pub fn is_array(&self) -> bool {
+        matches!(self.data, ObjectData::Array)
+    }
+
+    #[inline]
+    pub fn as_array(&self) -> Option<()> {
+        match self.data {
+            ObjectData::Array => Some(()),
+            _ => None,
         }
     }
-}
 
-/// Defines the different types of objects.
-#[derive(Finalize, Debug, Copy, Clone, Eq, PartialEq)]
-pub enum ObjectKind {
-    Function,
-    Array,
-    String,
-    Symbol,
-    Error,
-    Ordinary,
-    Boolean,
-    Number,
-    BigInt,
-}
+    /// Checks if it a `String` object.
+    #[inline]
+    pub fn is_string(&self) -> bool {
+        matches!(self.data, ObjectData::String(_))
+    }
 
-impl Display for ObjectKind {
-    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
-        write!(
-            f,
-            "{}",
-            match self {
-                Self::Function => "Function",
-                Self::Array => "Array",
-                Self::String => "String",
-                Self::Symbol => "Symbol",
-                Self::Error => "Error",
-                Self::Ordinary => "Ordinary",
-                Self::Boolean => "Boolean",
-                Self::Number => "Number",
-                Self::BigInt => "BigInt",
-            }
-        )
+    #[inline]
+    pub fn as_string(&self) -> Option<&String> {
```

Comment:
I think we should use `&str` here, instead of `&String`. This abstracts over the underlying implementation, that can be easily swapped by a `Box<str>` in the future. For that, we need to use `string.as_str()` later.